### ISSUE TITLE
iOS Safari bug

### ISSUE THREAD

**Issue by skrichten** [martinlaxenaire/curtainsjs#18]
The positions of the planes get out of sync on iOS Safari.  Chrome looks good though.

**Comment by martinlaxenaire** [comment]
Hi @skrichten,

could you provide more informations on your test device, iOS and Safari version?
Just tried with an iPhone 7, iOS 12.1.4 (should be Safari 12 if I understand well) and it's working well.

I don't personnally own any iOS device at the moment so debugging on such devices is kind of difficult for now.
I'd appreciate if anyone here has any additionnal input on this issue!

Thanks,

**Comment by skrichten** [comment]
Sorry for the lack of detail.  It was late and I was on my way to bed :)
iOS 12.4 on iPhone 8.  The positions don't always update properly when you scroll.  I can reproduce the issue with of the example files too.  https://www.curtainsjs.com/examples/multiple-planes/index.html

**Comment by martinlaxenaire** [comment]
Hey @skrichten,

I was able to reproduce the bug on the multiple planes example with a iPhone 7 and iOs 12.1.4.
It is therefore confirmed for **Safari version 12+ on iOS devices**.
This should be checked with previous iOS and Safari versions.

Something's definetely wrong with the `updatePosition()` method here. I'll try to debug that whenever I could.

Thanks!

**Comment by martinlaxenaire** [comment]
I'm suspecting that has something to do with the way iOS Safari handles the scroll event (based on that comment: https://github.com/w3c/IntersectionObserver/issues/315#issuecomment-449224717).

I've put up a codepen similar to the multiple planes example where I update the planes positions directly inside their render loop instead of inside the scroll event to test this scenario: https://codepen.io/martinlaxenaire/full/oNvBMeO

Will try it as soon as I can get my hand on an iPhone.

**Comment by skrichten** [comment]
I just had a quick look.  I don't see the image distortion happening on iOS.

**Comment by martinlaxenaire** [comment]
The distortion uniform is still calculated inside the scroll event, I did not change that. This pen purpose was just to see if the bug is coming from the library of from the Safari iOS scroll event itself.

Do you still got sync issues with that pen?

**Comment by skrichten** [comment]
on iOS Safari, I don't see any distortion, so I'm not sure if I am seeing the webGL planes or the original DOM images.

**Comment by martinlaxenaire** [comment]
Oh, I see. I've updated the codepen to always show a bit of distortion.
Thanks for your patience!

**Comment by skrichten** [comment]
The positions look good, but the distortion is not changing when I scroll... is it expected to? (it does on desktop)

**Comment by skrichten** [comment]
FYI, In my case, I am calling plane.updatePosition() in each plane's onRender.  When I have time I will setup a demo.

**Comment by martinlaxenaire** [comment]
Hey @skrichten thanks for your feedbacks!

That confirms what I was suspecting.

Good news is it has nothing to do with the library in itself!
Bad news is the scroll event is not reliable on iOS devices with Safari. A way to fix it would be to use the touchmove event on such devices but that means device and browser sniffing.

As for updating the planes positions inside their onRender callbacks with `updatePosition()`, this might not be a so good idea as it is triggering a layout repaint for each plane at each request animation frame call and can cause jankiness. See this post for further explanation and possible ways to avoid that: https://codepen.io/martinlaxenaire/post/webgl-enhanced-drag-slider-tutorial-with-curtains-js-part-3

**Comment by skrichten** [comment]
yeah, i was only calling updateposition that way as a quick test.

As for the bug, I'm not 100% sure about your conclusion.  I played with it a bit more and it actually works well enough until the screen size changes due to the browser chrome appearing or disappearing.

**Comment by skrichten** [comment]
https://three-dee.netlify.com/distortion
https://github.com/skrichten/ThreeDeeTalks/blob/master/src/components/CurtainsImage.js

I have the DOM element visible for the fourth image so you can see when things get out of sync.  Scrolling itself works fine, but when the browser bars appear or disappear, that's when they get out of sync.  So this seems like resize issue, rather than a scroll issue.

**Comment by martinlaxenaire** [comment]
Thanks @skrichten for the clarifications.

This looks like a confirmed bug (again).

Possible leads for debugging:
https://github.com/iamceege/tooltipster/issues/580
https://www.npmjs.com/package/ios-inner-height

I'll work on that bug next week as soon as I can get my hand on an Iphone.

**Comment by skrichten** [comment]
Let me know if you need some help testing this.

**Comment by martinlaxenaire** [comment]
Hi @skrichten,

The bug was indeed due to the Safari iOS navigation bar. The fixed canvas `getBoundingClientRect` returns a top property negative value as soon as the navigation bar hides, which is incorrect.

This should be fixed with the next release that will be shiped somewhere around last week of september I think.

Thanks for your help!

**Comment by martinlaxenaire** [comment]
Should be fixed in v4.0
